src/kaltura-player.js

```diff
@@ -535,6 +587,82 @@ class KalturaPlayer extends FakeEventTarget {
     return this._localPlayer.Error;
   }
 
+  _onChangeSourceStarted(): void {
+    this._configureOrLoadPlugins(this._pluginsConfig);
+    this._pluginManager.loadMedia();
+    this._reset = false;
+  }
+
+  _onEnded(): void {
+    // Make sure the all ENDED listeners have been invoked
+    setTimeout(() => {
+      if (this._adsController && !this._adsController.allAdsCompleted) {
+        this._eventManager.listenOnce(this._adsController, AdEventType.ALL_ADS_COMPLETED, () => {
+          this.dispatchEvent(new FakeEvent(CoreEventType.PLAYBACK_ENDED));
+        });
+      } else {
+        this.dispatchEvent(new FakeEvent(CoreEventType.PLAYBACK_ENDED));
+      }
+    });
+  }
+
+  _configureOrLoadPlugins(plugins: Object = {}): void {
+    const middlewares = [];
+    const uiComponents = [];
+    Object.keys(plugins).forEach(name => {
+      // If the plugin is already exists in the registry we are updating his config
+      const plugin = this._pluginManager.get(name);
+      if (plugin) {
+        plugin.updateConfig(plugins[name]);
+        plugins[name] = plugin.getConfig();
+      } else {
+        // We allow to load plugins as long as the player has no engine
+        if (!this._sourceSelected) {
+          try {
+            this._pluginManager.load(name, this, plugins[name]);
+          } catch (error) {
+            //bounce the plugin load error up
+            this.dispatchEvent(new FakeEvent(Error.Code.ERROR, error));
+          }
+          let plugin = this._pluginManager.get(name);
+          if (plugin) {
+            plugins[name] = plugin.getConfig();
+            if (typeof plugin.getMiddlewareImpl === 'function') {
+              // push the bumper middleware to the end, to play the bumper right before the content
+              plugin.name === 'bumper' ? middlewares.push(plugin.getMiddlewareImpl()) : middlewares.unshift(plugin.getMiddlewareImpl());
+            }
+
+            if (typeof plugin.getUIComponents === 'function') {
+              uiComponents.push(...(plugin.getUIComponents() || []));
+            }
+
+            if (typeof plugin.getEngineDecorator === 'function') {
+              registerEngineDecoratorProvider(plugin);
+            }
+          }
+        } else {
+          delete plugins[name];
+        }
+      }
+    });
+    this._pluginsUiComponents = uiComponents;
+    middlewares.forEach(middleware => this._localPlayer.playbackMiddleware.use(middleware));
+    Utils.Object.mergeDeep(this._pluginsConfig, plugins);
+    this._maybeCreateAdsController();
```

Comment:
why does this locate here?

Reply:
after the plugin creation or update, it's the time to create the ads controller which uses the plugin manager